### ISSUE TITLE
Panda renders shadows for lights that are out of view

### ISSUE THREAD

**Issue by rdb** [panda3d/panda3d#560]
It's not hard to [work around](https://discourse.panda3d.org/t/sample-using-directional-lights-shadows-effectively/24424) it in the application, but rather inconvenient if you have a large world with many lights.  Panda should do this automatically, by checking whether the shadow caster's frustum intersects with any of the scene camera view frusta before rendering shadows from that source.

Implementation-wise this is rather tricky because of how shadows are implemented, because the shadow lights are culled and rendered first, before we are actually performing culling for the main camera.  Also, there may be multiple cameras rendering different angles of the scene, not just the main camera.  Perhaps we need some kind of early pass over the windows where we collect all their SceneSetups and figure out which ones need to be rendered.

In the long term, this could be addressed with some kind of dependency graph for render passes, but it would be nice to have a solution for this in the short term.

**Comment by JoelStienlet** [comment]
For the "long term": can the shadow frustums and the PandaNodes be inserted into an octree? Then a list of any visible item could be computed for each camera by querying that octree.

**Comment by rdb** [comment]
I'm not sure I understand the technique you are describing specifically

What I'll do for now is that I'll change Panda's cull pass to save shadow passes for last; and when culling a regular scene pass, add its bounding volume to a list.  Then Panda will iterate over the shadow passes and only process a shadow pass if it overlaps with one of the bounding volumes.  (It doesn't really matter in which order things get culled anyway, as long as the shadow passes are *drawn* first.)

**Comment by JoelStienlet** [comment]
Yes the subject is certainly tricky enough to require some literature review before starting plans to address the issue for a long term solution.
If I understand correctly, this optimization will be most effective when there are multiple directional light sources in a scene, but only a few have a light cone that is visible from a camera?
And the less favourable case would be a scene with multiple visible objects, of which only a few are illuminated by a directional light source, because in that later case the light cones would be in the field of view of the camera, and thus need a shadow pass.

**Comment by rdb** [comment]
Yes, imagine a world where you have 100 street lights all casting shadows, but you are only within the effect radius of about 3 at a time.  The goal here is for Panda to automatically be clever enough to know not to render the shadow maps of the other 97 every frame.

For point lights, since Panda renders the faces in six passes, it is helpful to do this for every one of the sixth directions, so that when the point light is off screen Panda will only render the sides pointing towards the camera frustum.

I have implemented this locally already and will be checking it in after more testing, though I do want a more general solution in the long term, as well as something that will do frustum fitting automatically (as shown in [this example](https://discourse.panda3d.org/t/sample-using-directional-lights-shadows-effectively/24424)), as well as light culling.
